### ISSUE TITLE
Direct buffer count not decremented with HttpPostMultipartRequestDecoder

### ISSUE THREAD

**Issue by jameskleeh** [netty/netty#11174]
### Expected behavior
After an upload where the file was released, the DIRECT_MEMORY_COUNTER in PlatformDependent should not maintain a value for the file

### Actual behavior

After every upload the DIRECT_MEMORY_COUNTER retains a value for the previous upload. Eventually this will lead to an OutOfDirectMemoryError

Note that the file itself is in getting released correctly. I think this is related to how the `unencodedChunk` in `HttpPostMultipartRequestDecoder` is being used.

### Steps to reproduce

- `./gradlew run`
- Upload a 25-30mb file 5-10 times (the project README has an example curl)
- Observe the OutOfDirectMemoryError

You can also debug the run and set a breakpoint in PlatformDependent to observe the DIRECT_MEMORY_COUNTER value after uploading a file the first time and see it is not 0, or close to 0.

### Minimal yet complete reproducer code (or URL to code)

https://github.com/jameskleeh/netty-memory-leak

### Netty version

This started happening in 4.1.50 and is present in 4.1.63

### JVM version (e.g. `java -version`)
JDK 8

### OS version (e.g. `uname -a`)
Darwin MacBook-Pro.local 20.3.0 Darwin Kernel Version 20.3.0: Thu Jan 21 00:07:06 PST 2021; root:xnu-7195.81.3~1/RELEASE_X86_64 x86_64

**Comment by fredericBregier** [comment]
There are some benchmarks within Netty that shows exactly the contrary (whatever Memory, Mixed or Disk Factory based).
The results are not using big file however, just 512KB but with more than 6 000 upload/s, so very quick, and about 3 GB/s.
I've tested by changing size to 25 MB, increasing Factory size to 1GB, and it gives 131 upload/s, so 3.2 GB/s (so less than 1s to upload).

Trace of benchmark using this 25 MB file test: (HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderBigSimpleLevel is the reference in your cases)

     Benchmark                                                                           Mode  Cnt  Score   Error   Units
     HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderBigAdvancedLevel   thrpt    6  0,132 ± 0,005  ops/ms
     HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderBigDisabledLevel   thrpt    6  0,131 ± 0,005  ops/ms
     HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderBigParanoidLevel   thrpt    6  0,118 ± 0,003  ops/ms
     HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderBigSimpleLevel     thrpt    6  0,131 ± 0,001  ops/ms
     HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderHighAdvancedLevel  thrpt    6  0,124 ± 0,004  ops/ms
     HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderHighDisabledLevel  thrpt    6  0,125 ± 0,005  ops/ms
     HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderHighParanoidLevel  thrpt    6  0,075 ± 0,002  ops/ms
     HttpPostMultipartRequestDecoderBenchmark.multipartRequestDecoderHighSimpleLevel    thrpt    6  0,125 ± 0,004  ops/ms
     
     [817.169s][info   ][gc,heap,exit ] Heap
     [817.169s][info   ][gc,heap,exit ]  garbage-first heap   total 522240K, used 69233K [0x0000000602400000, 0x0000000800000000)
     [817.169s][info   ][gc,heap,exit ]   region size 2048K, 35 young (71680K), 2 survivors (4096K)
     [817.169s][info   ][gc,heap,exit ]  Metaspace       used 17320K, capacity 17729K, committed 17792K, reserved 1064960K
     [817.169s][info   ][gc,heap,exit ]   class space    used 1790K, capacity 1906K, committed 1920K, reserved 1048576K

I also test using the Junit tests testBIgFileUploadDelimiterInMiddleChunkDecoderXxxxFactory (where Xxx can be Disk, Memory, Mixed) using a 25 MB file, and it gives less than 0.5s (even 0,05s with Memory Factory) and only 1 MB of memory allocated bu the Factory.

Maybe you are using a Factory Mixe or Memory based (mixed with a too high limit than file size), such that it depends too much on memory ?

Also, have you clean the factory and HttpData correctly ?

      for (InterfaceHttpData httpData: decoder.getBodyHttpDatas()) {
          httpData.release();
          factory.removeHttpDataFromClean(request, httpData);
      }
      factory.cleanAllHttpData();
      decoder.destroy();

Regarding the other issue, it seems the issue is on your side.

Could you setup a simple test without your Micronaut code ?

**Comment by jameskleeh** [comment]
Yes all http data is getting released from the factory/decoder

**Comment by jameskleeh** [comment]
@fredericBregier The way you have shown is not the only way. Doing so that way does not work for us because some of the data may have already been released by the time we do cleanup. I have verified all of the data produced by the decoder that is our responsibility to release has been released.

You can choose to ignore the issue, but it definitely is an issue. The Junit tests are obviously not catching this case

**Comment by fredericBregier** [comment]
I didn't say there is no issue, I said that the way to free correctly the buffers is this way.
The reason is the difference of handling buffers between in Memory and on Disk.
You can check every data before calling the clean up using `buffer.refCnt() > 0` to ensure it is ok to do so.
I didn't choose to ignore the issue, I just say that we have made Junit tests, and they didn't show this issue.
So the reason we'r (or I'm) asking you a simple reproducer in order to be able to detect, add the Junit test and fix it, if is real.

**Comment by fredericBregier** [comment]
Javadoc for Factory:

    /**
     * Remove all InterfaceHttpData from virtual File storage from clean list for all requests
     */
    void cleanAllHttpData();

It means that only "file" based are released there, not "memory" based one.

Then on Decoder:

    /**
     * Destroy the {@link HttpPostMultipartRequestDecoder} and release all it resources. After this method
     * was called it is not possible to operate on it anymore.
     */
    @Override
    public void destroy() {
        // Release all data items, including those not yet pulled
        cleanFiles();

        destroyed = true;

        if (undecodedChunk != null && undecodedChunk.refCnt() > 0) {
            undecodedChunk.release();
            undecodedChunk = null;
        }
    }

Except that it calls `cleanFiles()` which calls ` factory.cleanRequestHttpData(request);` which Javadoc says:

    /**
     * Remove all InterfaceHttpData from virtual File storage from clean list for the request
     *
     * @param request associated request
     */
    void cleanRequestHttpData(HttpRequest request);

So yes, there is an issue, since only "Disk" based are free.
So what I can propose is to change as the following:


    /**
     * Destroy the {@link HttpPostMultipartRequestDecoder} and release all it resources. After this method
     * was called it is not possible to operate on it anymore.
     */
    @Override
    public void destroy() {
        // Release all data items, including those not yet pulled
        cleanFiles();
       
        // Explicitely remove all HttpData but only if refCnt is > 0
        for (InterfaceHttpData httpData: getBodyHttpDatas()) {
           if (httpData.refCnt > 0) {
               httpData.release();
          }
          factory.removeHttpDataFromClean(request, httpData);
        }
        destroyed = true;

        if (undecodedChunk != null && undecodedChunk.refCnt() > 0) {
            undecodedChunk.release();
            undecodedChunk = null;
        }
    }

**Comment by jameskleeh** [comment]
You can do that however it has no relation to this issue. As I've said several times all of the data produced by the factory is getting cleaned correctly in this example.

**Comment by fredericBregier** [comment]
For your intention, this issue was introduced 2 years ago (not by the last changes).

**Comment by fredericBregier** [comment]
@jameskleeh Yes, I mixed up the 2 issues. You're right. Sorry for that.

Now, as I don't have this issue (and the change instead improves quite a lot, both on memory and time), I can't figure out without a simple reproducer...

**Comment by jameskleeh** [comment]
@fredericBregier Is there any option for commercial support for Netty?

**Comment by fredericBregier** [comment]
@jameskleeh OK, then it was the right place for my comment (except that the other user referenced in your other issue about performance his own issue on memory leak).
You have a memory leak, and I explain to you why.

I will soon propose a patch to get back to 2 years ago feature, which I was not part of.
I've changed a Junit to reflect the correct behavior now when using simply `decoder.destroy()`

As for commercial support, I don't know. I'm a benevol particpating coder.

**Comment by fredericBregier** [comment]
Well, we can see later on (too late for me right now).
When calling `decoder.destroy()`, the undecodedChunk is released.
I found out maybe another point (extra commit done and pushed), but it shouldn't be the case.
However, this is the only point where I can figure out a position where the undecodedChunk is not released correctly.

Can you check when you have time?

**Comment by jameskleeh** [comment]
@fredericBregier I tried to reproduce the memory leak with 4.1.63 and it seems I can no longer reproduce it. Given how long it takes to send each request I was debugging the state of PlatformDependent and did observe the memory counter going up for each request. I was sure I also observed the error however @doom369 could not reproduce the memory leak in `4.1.63` so I tried it again and I can't reproduce. This definitely is an issue in `4.1.59` and it seems like something was changed to fix it between `4.1.60` and `4.1.63`. I appreciate you helping out and sorry for wasting your time with this specific issue.
